lightningd/invoice.c

```diff
@@ -167,6 +168,49 @@ struct invoice_payment_hook_payload {
 	/* FIXME: Include raw payload! */
 };
 
+#ifdef DEVELOPER
+static void invoice_payment_add_tlvs(struct json_stream *stream,
+				     struct htlc_set *hset)
+{
+	struct htlc_in *hin;
+	struct tlv_tlv_payload *tlvs;
+	u64 prevtype = 0;
+	assert(tal_count(hset->htlcs) > 0);
+
+	/* Pick the first HTLC as representative for the entire set. */
+	hin = hset->htlcs[0];
+
+	if (hin->payload->type != ONION_TLV_PAYLOAD)
+		return;
+	tlvs = hin->payload->tlv;
+
+	json_object_start(stream, "extra_tlvs");
+
+	for (size_t i = 0; i < tal_count(tlvs->fields); i++) {
+		struct tlv_field *field = &tlvs->fields[i];
+		/* If we have metadata attached it is not an extra TLV field. */
+		if (field->meta == NULL) {
+			/* If this is a new type, and we have an open one, close
+			 * the old one before opening a new one. */
+			if (prevtype != field->numtype && prevtype != 0)
+				json_array_end(stream);
+			/* Now, if it is a new type, open a new array for this type*/
+			if (prevtype != field->numtype)
+				json_array_start(stream,
+						 tal_fmt(stream, "%" PRIu64,
+							 field->numtype));
+			/* Finally add this element. */
+			json_add_hex_talarr(stream, NULL, field->value);
+			prevtype = field->numtype;
+		}
+	}
+	/* If we opened an array at all, now is the time to close it. */
+	if (prevtype != 0)
+		json_array_end(stream);
+	json_object_end(stream);
```

Comment:
BTW, JSON schemas has taught me that variable member names are an anti-pattern.  So make extra_tlvs an array of objects, with each object "type": u64, "data": hex?  Code will be simpler too...

Reply:
The main issue here is that the spec technically allows a field to be present multiple times, hence the TLV-type-to-array-of-values indirection. More than happy not to support this :-)

Reply:
Sure, then you just have duplicate fields:

```
[ {"type": 8, "data": "001122"}, {"type": 8, "data": "00112233"} ]
```

Reply:
Actually, they *can* but it's invalid:

```
 - MUST order `tlv_record`s in a `tlv_stream` by strictly-increasing `type`,
   hence MUST not produce more than a single TLV record with the same `type`
```

And:

```
 - if decoded `type`s are not strictly-increasing (including situations when
   two or more occurrences of the same `type` are met):
   - MUST fail to parse the `tlv_stream`.
```

Reply:
Right, I was going by the validation code, should have checked the spec itself :-)

Reply:
Will amend :+1: